File: app/src/main/java/com/iug/coursehub/feature/dashboard/courses_screen/CoursesFragment.java
package com.iug.coursehub.feature.dashboard.courses_screen;

import static androidx.navigation.Navigation.findNavController;

import android.content.DialogInterface;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.iug.coursehub.R;
import com.iug.coursehub.data.local.db.CoursesRepository;
import com.iug.coursehub.data.local.db.entity.Course;
import com.iug.coursehub.databinding.FragmentCoursesBinding;
import com.iug.coursehub.feature.dashboard.lessons_screen.LessonsFragment;

import java.util.List;

public class CoursesFragment extends Fragment {

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        FragmentCoursesBinding binding = FragmentCoursesBinding.inflate(inflater, container, false);

        CoursesViewModel viewModel = new ViewModelProvider(this).get(CoursesViewModel.class);

        binding.rvCourses.setLayoutManager(new LinearLayoutManager(requireContext()));
        LiveData<List<Course>> allCoursesLiveData = CoursesRepository.getAllCourses();
        allCoursesLiveData.observe(getViewLifecycleOwner(), new Observer<List<Course>>() {
            @Override
            public void onChanged(List<Course> allCourses) {
                if (allCourses.isEmpty()) {
                    binding.content.setVisibility(View.GONE);
                    binding.sorryMessage.setVisibility(View.VISIBLE);
                } else {
                    binding.content.setVisibility(View.VISIBLE);
                    binding.sorryMessage.setVisibility(View.GONE);

                    CoursesAdapter adapter = new CoursesAdapter(
                            allCourses,
                            (id) -> {
                                Bundle bundle = new Bundle();
                                bundle.putInt(LessonsFragment.ARG_COURSE_ID, id);

                                findNavController(requireView()).navigate(
                                        R.id.action_coursesFragment_to_lessonsFragment,
                                        bundle
                                );
                            },
                            (id) -> {
                                Bundle bundle = new Bundle();
                                bundle.putInt(LessonsFragment.ARG_COURSE_ID, id);
                                findNavController(requireView()).navigate(
                                        R.id.action_coursesFragment_to_addEditCourseFragment,
                                        bundle
                                );
                            },
                            (course) -> {
                                new MaterialAlertDialogBuilder(requireContext())
                                        .setIcon(R.drawable.ic_delete)
                                        .setTitle("Deleting Course")
                                        .setMessage("Are you sure you want to delete this course?")
                                        .setPositiveButton("Yes", new DialogInterface.OnClickListener() {
                                            @Override
                                            public void onClick(DialogInterface dialog, int which) {
                                                viewModel.deleteCourse(course);
                                                dialog.dismiss();
                                            }
                                        }).setNegativeButton("No", new DialogInterface.OnClickListener() {
                                            @Override
                                            public void onClick(DialogInterface dialog, int which) {
                                                dialog.dismiss();
                                            }
                                        }).create()
                                        .show();
                            }
                    );
                    binding.rvCourses.setAdapter(adapter);
                }
            }
        });

        return binding.getRoot();
    }
}

File: app/src/main/java/com/iug/coursehub/feature/dashboard/courses_screen/CoursesFragment.kt
package com.iug.coursehub.feature.dashboard.courses_screen

import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.iug.coursehub.R
import com.iug.coursehub.data.local.db.CoursesRepository
import com.iug.coursehub.databinding.FragmentCoursesBinding
import com.iug.coursehub.feature.dashboard.DashboardNavHostFragment
import com.iug.coursehub.feature.dashboard.add_edit_course_screen.AddEditCourseFragment
import com.iug.coursehub.feature.dashboard.lessons_screen.LessonsFragment

class CoursesFragment : Fragment() {

    private val viewModel by viewModels<CoursesViewModel>()

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val binding = FragmentCoursesBinding.inflate(inflater, container, false)

        binding.apply {
            rvCourses.apply {
                layoutManager = LinearLayoutManager(requireContext())

                val allCoursesLiveData = CoursesRepository.getAllCourses()
                allCoursesLiveData.observe(viewLifecycleOwner) { allCourses ->
                    if (allCourses.isEmpty()) {
                        content.visibility = View.GONE
                        sorryMessage.visibility = View.VISIBLE
                    } else {
                        content.visibility = View.VISIBLE
                        sorryMessage.visibility = View.GONE

                        val adapter = CoursesAdapter(
                            courses = allCourses,
                            onShowLessonsClick = {
                                findNavController().navigate(
                                    R.id.action_coursesFragment_to_lessonsFragment,
                                    Bundle().apply {
                                        putInt(LessonsFragment.ARG_COURSE_ID, it)
                                    }
                                )
                            },
                            onEditCourseClick = {
                                findNavController().navigate(
                                    R.id.action_coursesFragment_to_addEditCourseFragment,
                                    Bundle().apply {
                                        putInt(AddEditCourseFragment.ARG_COURSE_ID, it)
                                    }
                                )
                            },
                            onDeleteCourseClick = {
                                MaterialAlertDialogBuilder(requireContext())
                                    .setIcon(R.drawable.ic_delete)
                                    .setTitle("Deleting Course")
                                    .setMessage("Are you sure you want to delete this course?")
                                    .setPositiveButton("Yes") { dialogInterface, _ ->
                                        viewModel.deleteCourse(it)
                                        dialogInterface.dismiss()
                                    }.setNegativeButton("No") { dialogInterface, _ ->
                                        dialogInterface.dismiss()
                                    }.create()
                                    .show()
                            }
                        )
                        this.adapter = adapter
                    }
                }
            }

            fabAddCourse.setOnClickListener {
                findNavController().navigate(R.id.action_coursesFragment_to_addEditCourseFragment)
            }

            btnSignOut.setOnClickListener {
                viewModel.signOut()
            }
        }

        viewModel.uiAction.observe(viewLifecycleOwner) { action ->
            when(action) {
                CoursesViewModel.UiAction.NavigateToAuthScreen -> {
                    DashboardNavHostFragment.viewModel.navigateToAuthScreen()
                }
            }
        }

        return binding.root
    }
}